### PR TITLE
Initial Checkin

### PR DESCRIPTION
**Pull request by kfox1111**
This adds an initial implementation. It works
but has little error handling. Do not use for
production. It is based on the layout from
the csi-driver-host-path repo.

### CODE REVIEW

**Comment by kfox1111** [comment]
Most of the files here is from the host path driver. vendor and release-tools directories are unchanged from host path. Most of the new code is under cmd and pkg.

**Review by pohly — CHANGES_REQUESTED** [review]
`release-tools` must be added and updated via `git subtree`.

Please:
1. set up Travis CI
2. merge PR #3 
3. rebase and merge this PR

**Comment by kfox1111** [comment]
Ok. Thanks for the help. Still learning how all this stuff works. :)

**Comment by kfox1111** [comment]
Rebased, lint errors fixed, test stubs put in place.

File: deploy/deploy-image.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+BASE_DIR=$(dirname "$0")
+K8S_RELEASE=${K8S_RELEASE:-"release-1.13"}
+INSTALL_CRD=${INSTALL_CRD:-"false"}
+
+# apply CSIDriver and CSINodeInfo API objects
+if [[ "${INSTALL_CRD}" =~ ^(y|Y|yes|true)$ ]] ; then
+    echo "installing CRDs"
+    kubectl apply -f https://raw.githubusercontent.com/kubernetes/csi-api/${K8S_RELEASE}/pkg/crd/manifests/csidriver.yaml --validate=false
+    kubectl apply -f https://raw.githubusercontent.com/kubernetes/csi-api/${K8S_RELEASE}/pkg/crd/manifests/csinodeinfo.yaml --validate=false
+fi
```
Comment (saad-ali) [inline deploy/deploy-image.sh]: This is no longer required as of k8s1.13: https://kubernetes-csi.github.io/docs/csi-objects.html

File: deploy/deploy-image.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+# done here and referenced elsewhere.
+
+# The script assumes that kubectl is available on the OS path 
+# where it is executed.
+
+set -e
+set -o pipefail
+
+BASE_DIR=$(dirname "$0")
+K8S_RELEASE=${K8S_RELEASE:-"release-1.13"}
```
Comment (msau42) [inline deploy/deploy-image.sh]: Does this actually work on 1.13?

File: deploy/image/csi-image-plugin.yaml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  template:
+    metadata:
+      labels:
+        app: csi-imageplugin
+    spec:
+      hostNetwork: true
+      containers:
+        - name: node-driver-registrar
+          image: quay.io/k8scsi/csi-node-driver-registrar:v1.0.2
+          imagePullPolicy: Always
```
Comment (msau42) [inline deploy/image/csi-image-plugin.yaml]: is this necessary?

File: pkg/image/driver.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+func NewDriver(nodeID, endpoint string) *driver {
+	glog.Infof("Driver: %v version: %v", driverName, version)
+
+	d := &driver{}
+
+	d.endpoint = endpoint
+
+	csiDriver := csicommon.NewCSIDriver(driverName, version, nodeID)
+	csiDriver.AddVolumeCapabilityAccessModes([]csi.VolumeCapability_AccessMode_Mode{csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER})
+	// NFS plugin does not support ControllerServiceCapability now.
```
Comment (saad-ali) [inline pkg/image/driver.go]: NFS plugin?

File: deploy/image/csi-image-plugin.yaml [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                  apiVersion: v1
+                  fieldPath: spec.nodeName
+          volumeMounts:
+          - mountPath: /csi
+            name: socket-dir
+          - mountPath: /registration
+            name: registration-dir
+
+        - name: image
+          image: quay.io/k8scsi/imageplugin:v1.0.1
```
Comment (msau42) [inline deploy/image/csi-image-plugin.yaml]: Is this a valid version? should it be canary for now?

File: pkg/image/driver.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+func NewDriver(nodeID, endpoint string) *driver {
+	glog.Infof("Driver: %v version: %v", driverName, version)
+
+	d := &driver{}
+
+	d.endpoint = endpoint
+
+	csiDriver := csicommon.NewCSIDriver(driverName, version, nodeID)
+	csiDriver.AddVolumeCapabilityAccessModes([]csi.VolumeCapability_AccessMode_Mode{csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER})
```
Comment (msau42) [inline pkg/image/driver.go]: Is this correct?

File: pkg/image/controllerserver.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+import (
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"golang.org/x/net/context"
+
+	"github.com/kubernetes-csi/drivers/pkg/csi-common"
+)
+
+type controllerServer struct {
+	*csicommon.DefaultControllerServer
```
Comment (msau42) [inline pkg/image/controllerserver.go:27]: We're trying to deprecate and move away from csicommon

File: README.md [hunk 8]
```diff
@@ -14,3 +14,27 @@ You can reach the maintainers of this project at:
 ### Code of conduct
 
 Participation in the Kubernetes community is governed by the [Kubernetes Code of Conduct](code-of-conduct.md).
+
+
```
Comment (saad-ali) [inline README.md:23]: You should:

1. add brief description of how it works: e.g. this driver uses https://github.com/containers/buildah to mount a container image, etc.
2. Expand the usage: how do you deploy this CSI driver? Once it is deployed, how do you use this CSI driver?

File: pkg/image/nodeserver.go [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+	glog.V(4).Infof("target %v\nfstype %v\ndevice %v\nreadonly %v\nvolumeId %v\nattributes %v\n mountflags %v\n",
+		targetPath, fsType, deviceId, readOnly, volumeId, attrib, mountFlags)
+
+	options := []string{"bind"}
+	if readOnly {
+		options = append(options, "ro")
+	}
+
+	args := []string{"mount", volumeId}
+	ns.execPath = "/bin/buildah" // FIXME
```
Comment (msau42) [inline pkg/image/nodeserver.go:110]: what is this binary?

**Review by saad-ali — CHANGES_REQUESTED** [review]
/approve

@msau42 for LGTM

File: pkg/image/driver.go (hunk 4 above)
Comment (kfox1111) [reply pkg/image/driver.go]: I missed that comment. It started life out as the nfs plugin. I'll remove it.

File: deploy/deploy-image.sh (hunk 2 above)
Comment (kfox1111) [reply deploy/deploy-image.sh]: ah. no. it doesn't actually work in 1.14.[0-1] either...

File: deploy/image/csi-image-plugin.yaml (hunk 3 above)
Comment (kfox1111) [reply deploy/image/csi-image-plugin.yaml]: not sure I follow? The imagePullPolicy or the sidecar? I think the sidecar's needed isnt it?

File: deploy/image/csi-image-plugin.yaml (hunk 5 above)
Comment (kfox1111) [reply deploy/image/csi-image-plugin.yaml]: +1

What should the actual repo be too?

File: README.md [hunk 10]
```diff
@@ -25,7 +30,35 @@ Participation in the Kubernetes community is governed by the [Kubernetes Code of
...
+    image: nginx:1.13-alpine
+    ports:
+    - containerPort: 80
+    volumeMount:
+    - name: data
+      mountPath: /usr/share/nginx/html
+  volumes:
+  - name: data
+    csi:
+      driver: csi-image
```
Comment (msau42) [inline README.md]: I know we don't do this currently for the other sample drivers, but do we want to have a prefix, like "storage.sigs.k8s.io"?

File: deploy/deploy-image.sh (hunk 2 above)
Comment (msau42) [reply deploy/deploy-image.sh]: I guess it's not used anymore, so we can remove this variable?

File: pkg/image/driver.go (hunk 6 above)
Comment (kfox1111) [reply pkg/image/driver.go]: I kind of wondered about that... Its in a gray area in the spec.

It supports multiple nodes attaching to the ephemeral driver at the same time. So in one way it is multi attach.  But no data is shared across nodes.

File: pkg/image/controllerserver.go (hunk 7 above)
Comment (kfox1111) [reply pkg/image/controllerserver.go:27]: ok. I'd be happy to rewrite it if thats the way to go, but I really would like to get something merged to start testing against. Can this wait till a followup pr?

File: README.md [hunk 11]
```diff
@@ -15,6 +15,11 @@ You can reach the maintainers of this project at:
 
 Participation in the Kubernetes community is governed by the [Kubernetes Code of Conduct](code-of-conduct.md).
 
+## How it works:
+
+Currently the driver makes use of buildah to download the container image if it is not already available, launch a new instance of it named after the volumeHandle, and mount it.
+
+In the future, integration with CRI would be desirable so the driver could ask via CRI that the Container Runtime perform these activities in a generic way.
 
 ## Usage:
```
Comment (msau42) [inline README.md:25]: Maybe add a note somewhere here about the min K8s version (master or still under dev)

File: pkg/image/nodeserver.go (hunk 9 above)
Comment (kfox1111) [reply pkg/image/nodeserver.go:110]: https://github.com/containers/buildah

Ideally we would get CRI support added for this use case. For now, buildah supports the use case out of the box so useful to bootstrap the project.

File: README.md (hunk 10 above)
Comment (kfox1111) [reply README.md]: works for me. so, storage.sigs.k8s.io/image ? or image.storage.sigs.k8s.io?

File: README.md (hunk 11 above)
Comment (kfox1111) [reply README.md:25]: +1

File: README.md (hunk 10 above)
Comment (msau42) [reply README.md]: Oh actually driver name is not a prefix.  @saad-ali suggests "image.csi.k8s.io"

File: pkg/image/controllerserver.go (hunk 7 above)
Comment (msau42) [reply pkg/image/controllerserver.go:27]: Yeah followup is fine

File: pkg/image/driver.go (hunk 6 above)
Comment (msau42) [reply pkg/image/driver.go]: Yeah I think it's probably better to put single writer for now until we can fix the access modes issues

File: deploy/image/csi-image-plugin.yaml (hunk 5 above)
Comment (msau42) [reply deploy/image/csi-image-plugin.yaml]: I think the repo is fine. Maybe just canary as the tag

File: deploy/image/csi-image-plugin.yaml (hunk 3 above)
Comment (msau42) [reply deploy/image/csi-image-plugin.yaml]: i mean the imagePullPolicy: always

File: README.md (hunk 10 above)
Comment (kfox1111) [reply README.md]: Sounds good to me.

File: pkg/image/driver.go (hunk 6 above)
Comment (kfox1111) [reply pkg/image/driver.go]: ok.

File: cmd/imageplugin/main.go [hunk 12]
```diff
@@ -29,7 +29,7 @@ func init() {
 
 var (
 	endpoint   = flag.String("endpoint", "unix://tmp/csi.sock", "CSI endpoint")
-	driverName = flag.String("drivername", "csi-image", "name of the driver")
+	driverName = flag.String("drivername", "image.csi.k8s.io", "name of the driver")
```
Comment (msau42) [inline cmd/imageplugin/main.go:32]: does the driver name need to be configurable?

**Review by msau42 — COMMENTED** [review]
generally lgtm, please squash

File: deploy/image/csi-image-plugin.yaml (hunk 3 above)
Comment (kfox1111) [reply deploy/image/csi-image-plugin.yaml]: Ah. Yeah, I can update that to IfNotPresent

File: cmd/imageplugin/main.go (hunk 12 above)
Comment (kfox1111) [reply cmd/imageplugin/main.go:32]: It was in other drivers. I just kept it. If we think it should be hard coded, we can change it back.

**Comment by msau42** [comment]
lgtm, can you squash your commits first

**Comment by kfox1111** [comment]
> lgtm, can you squash your commits first

Squashed. Thanks for reviewing.

**Pull request merged**